Automatically populate `CONSUL_HTTP_ADDR` for connect native tasks in host networking mode.

apollo13 opened the pull request:
Fixes #10239. This only works if the `network` stanza is defined at the group level; otherwise
https://github.com/hashicorp/nomad/blob/3d9464023dbde64417468f4e623643d31fc3e7a5/nomad/job_endpoint_hook_connect.go#L100-L107
will skip the `groupConnectHook` which is responsible for setting the task kind to `connect-native`…

/cc @shoenig for a review :)

shoenig commented:
Hi @apollo13 , thanks for the PR!

Looks like a few `TestTaskRunner_ConnectNativeHook_*` tests are failing now, since they verify the expected environment which has now changed. We'll also need to make an entry in the [version-specific upgrade guide](https://www.nomadproject.io/docs/upgrade/upgrade-specific) ([source](https://github.com/hashicorp/nomad/blob/v1.0.4/website/content/docs/upgrade/upgrade-specific.mdx)) documenting what users should expect. I don't imagine the change will interrupt existing services, but since the environment is now setting a variable which was previously unset, we do need to highlight the change.

I think we'll also want to document the variable in https://www.nomadproject.io/docs/runtime/environment ([source](https://github.com/hashicorp/nomad/blob/v1.0.4/website/content/docs/runtime/environment.mdx)) which we missed in the original PR. How about in an new subheading called "Consul-related variables"?

apollo13 commented:
> Looks like a few `TestTaskRunner_ConnectNativeHook_*` tests are failing now, since they verify the expected environment which has now changed.

Fixed. Also added docs as requested, feel free to reword them as you see fit.

shoenig reviewed (approved):
LGTM, thanks @apollo13 !
